# Practice with R

In [1]:
library(ISLR)
attach(Auto)

## Validation set approach

### Resampling with different seeds

In [2]:
# Set the random seed with a constant to stabilize results of random processes:
set.seed (1)

# Generate random train indices:
(train = sample(392, 196))

[1] 324 167 129 299 270 187 307  85 277 362 330 263 329  79 213  37 105 217
 [19] 366 165 290 383  89 289 340 326 382  42 111  20  44 343  70 121  40 172
 [37]  25 248 198  39 298 280 160  14 130  45  22 206 230 193 104 367 255 341
 [55] 342 103 331  13 296 375 176 279 110  84  29 141 252 221 108 304  33 347
 [73] 149 287 102 145 118 323 107  64 224 337  51 325 372 138 390 389 282 143
 [91] 285 170  48 204 295  24 181 214 225 163  43   1 328  78 284 116 233  61
[109]  86 374  49 242 246 247 239 219 135 364 363 310  53 348  65 376 124  77
[127] 218  98 194  19  31 174 237  75  16 358   9  50  92 122 152 386 207 244
[145] 229 350 355 391 223 373 309 140 126 349 344 319 258  15 271 388 195 201
[163] 318  17 212 127 133  41 384 392 159 117  72  36 315 294 157 378 313 306
[181] 272 106 185  88 281 228 238 368  80  30  93 234 220 240 369 164

In [3]:
# Fit a linear regression model with horsepower to predict mpg on train set:
lm.fit=lm(mpg~horsepower,data=Auto,subset=train)

# mean((mpg-predict(lm.fit2,Auto))[-train]~2) # ---> This line used wrong syntax to calculate Test MSE.

# Instead, we use the following line to calculate MSE of our model on test set:
mean((mpg - predict(lm.fit,Auto))[-train]^2)

[1] 23.26601

In [4]:
# Fit a 2nd degree polynomial regression model with horsepower to predict mpg on train set:
lm.fit2 = lm(mpg~poly(horsepower, 2) ,data=Auto,subset = train)

# Calculate MSE on test set:
mean((mpg - predict(lm.fit2,Auto))[-train]^2)

[1] 18.71646

In [5]:
# Fit a 3rd degree polynomial regression model with horsepower to predict mpg on train set:
lm.fit3 = lm(mpg~poly(horsepower, 3) ,data=Auto,subset = train)

# Calculate MSE on test set:
mean((mpg - predict(lm.fit3,Auto))[-train]^2)

[1] 18.79401

In [6]:
# Repeat above steps with another seed value:
set.seed(2)
train = sample(392 ,196)

In [7]:
# Fit a linear regression model with horsepower to predict mpg on train set:
lm.fit=lm(mpg~horsepower,data=Auto,subset=train)

# Calculate MSE on test set:
mean((mpg - predict(lm.fit,Auto))[-train]^2)

[1] 25.72651

In [8]:
# Fit a 2nd degree polynomial regression model with horsepower to predict mpg on train set:
lm.fit2 = lm(mpg~poly(horsepower, 2) ,data=Auto,subset = train)

# Calculate MSE on test set:
mean((mpg - predict(lm.fit2,Auto))[-train]^2)

[1] 20.43036

In [9]:
# Fit a 3rd degree polynomial regression model with horsepower to predict mpg on train set:
lm.fit3 = lm(mpg~poly(horsepower, 3) ,data=Auto,subset = train)

# Calculate MSE on test set:
mean((mpg - predict(lm.fit3,Auto))[-train]^2)

[1] 20.38533

## Leave-One-Out Cross-Validation

In [10]:
# Fit a generalized linear model, which in this case is the same as fitting a linear regression model (lm.fit),
# on horsepower to predict mpg
glm.fit = glm(mpg~horsepower ,data= Auto)
coef(glm.fit)

(Intercept)  horsepower 
 39.9358610  -0.1578447

In [11]:
# Fit a linear regression model on horsepower to predict mpg
lm.fit=lm(mpg~horsepower,data=Auto)
coef(lm.fit)

(Intercept)  horsepower 
 39.9358610  -0.1578447

In [12]:
# Import boot library whicn contains cross-validation functions
library(boot)

# Fit a linear regression model on horsepower with glm.fit:
glm.fit = glm(mpg~horsepower ,data= Auto)

# Evaluate the model with K-Fold Cross-Validation:
cv.err = cv.glm(Auto, glm.fit)
# K isn't given, so cv.glm will use the default value of K, which is equal to the number of observations.
# When K is equal to the number of observations, K-Fold Cross-Validation becomes Leave-One-Out Cross-Validation

cv.err[['delta']]
# Delta is a vector of length two. 
# The first component is the raw cross-validation estimate of prediction error. 
# The second component is the adjusted cross-validation estimate.
# The adjustment is designed to compensate for the bias introduced by not using leave-one-out cross-validation.

[1] 24.23151 24.23114

In [13]:
# Create a vector of zeros with length 5:
cv.error = rep(0,5)

# Evaluate 5 polynomial regression models of 5 different degrees from 1st to 5th with Leave-One-Out Cross-Validation:
for (i in 1:5){
    # Fit polynomial model of ith degree on horsepower to predict mpg:
    glm.fit=glm(mpg~poly(horsepower,i),data=Auto)
    # Calculate Leave-One-Out Cross-Validation error of ith model:
    cv.error[i] = cv.glm(Auto,glm.fit)$delta[1]
}
cv.error

[1] 24.23151 19.24821 19.33498 19.42443 19.03321

## K-Fold Cross-Validation

In [14]:
set.seed(17)
# Create a vector of zeros with length 10:
cv.error.10 = rep(0 ,10)

# Evaluate 10 polynomial regression models of 10 different degrees from 1st to 10th with K-Fold Cross-Validation:
for (i in 1:10) {
    # Fit polynomial model of ith degree on horsepower to predict mpg:
    glm.fit=glm(mpg~poly(horsepower,i),data=Auto)
    # Calculate 10-Fold Cross-Validation error of ith model:
    cv.error.10[i]=cv.glm(Auto, glm.fit, K=10)$delta[1]
}
cv.error.10

[1] 24.27207 19.26909 19.34805 19.29496 19.03198 18.89781 19.12061 19.14666
 [9] 18.87013 20.95520

## The Bootstrap
### Estimating the Accuracy of a Statistic of Interest

In [15]:
# Write a function to calculate the alpha value with the given formula, which is an estimation of investment risk:
alpha.fn=function(data,index){
  X=data$X[index]
  Y=data$Y[index]
  return((var(Y)-cov(X,Y))/(var(X)+var(Y)-2*cov(X,Y)))
}

In [16]:
# Calculate alpha for the first 100 observations:
alpha.fn(Portfolio, 1:100)

[1] 0.5758321

In [17]:
# The next command uses the sample() function to randomly select
# 100 observations from the range 1 to 100, with replacement.
# This is equivalent to constructing a new bootstrap data set and
# recomputing alpha based on the new data set.

set.seed(1)
alpha.fn(Portfolio, sample(100,100,replace=T))

[1] 0.7368375

In [18]:
# Produce R=1000 bootstrap estimates for alpha using boot()
boot(Portfolio, alpha.fn, R=1000)


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Portfolio, statistic = alpha.fn, R = 1000)


Bootstrap Statistics :
     original       bias    std. error
t1* 0.5758321 -0.001695873  0.09366347

The final output shows that using the original data, alpha = 0.5758,
and that the bootstrap estimate for SE(a) is 0.0936

###  Estimating the Accuracy of a Linear Regression Model

In [19]:
# Fit a linear regression model on a sample set with given index:
boot.fn=function(data,index) {
    return(coef(lm(mpg~horsepower,data=data,subset=index)))
}
boot.fn(Auto, 1:392)

(Intercept)  horsepower 
 39.9358610  -0.1578447

In [20]:
set.seed (1)
# Fit a linear regression model on a bootstrap sample:
boot.fn(Auto,sample(392,392,replace=T))

(Intercept)  horsepower 
 40.3404517  -0.1634868

In [21]:
# Fit a linear regression model on another bootstrap sample:
boot.fn(Auto,sample(392,392,replace=T))

(Intercept)  horsepower 
 40.1186906  -0.1577063

In [22]:
# Produce R=1000 bootstrap estimates for Linear Regression model using boot()
boot(Auto,boot.fn,1000)


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Auto, statistic = boot.fn, R = 1000)


Bootstrap Statistics :
      original        bias    std. error
t1* 39.9358610  0.0544513229 0.841289790
t2* -0.1578447 -0.0006170901 0.007343073

In [23]:
# Get summary of linear model coef:
summary(lm(mpg~horsepower,data=Auto))$coef

Estimate   Std. Error  t value   Pr(>|t|)     
(Intercept) 39.9358610 0.717498656  55.65984 1.220362e-187
horsepower  -0.1578447 0.006445501 -24.48914  7.031989e-81

In [24]:
# Produce bootstrap estimation of 2nd degree polynomial regression model:
boot.fn=function(data,index) {
    coefficients(lm(mpg~horsepower+I(horsepower^2),data=data, subset = index))
}
set.seed(1)
boot(Auto,boot.fn ,1000)


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Auto, statistic = boot.fn, R = 1000)


Bootstrap Statistics :
        original        bias     std. error
t1* 56.900099702  3.511640e-02 2.0300222526
t2* -0.466189630 -7.080834e-04 0.0324241984
t3*  0.001230536  2.840324e-06 0.0001172164

# Exercise 1
In Chapter 4, we used logisitc regression to predict the probability of “default” using “income” and “balance” on the “Default” data set. We will now estimate the test error of this logistic regression model using the validation set approach. Do not forget to set a random seed before beginning your analysis.

In [25]:
library(ISLR)
attach(Default)
set.seed(1)
head(Default)

default student balance   income   
1 No      No       729.5265 44361.625
2 No      Yes      817.1804 12106.135
3 No      No      1073.5492 31767.139
4 No      No       529.2506 35704.494
5 No      No       785.6559 38463.496
6 No      Yes      919.5885  7491.559

## Task (a)
Fit a logistic regression model that uses “income” and “balance” to predict “default”.

In [26]:
fit.glm = glm(default ~ income + balance, data = Default, family = "binomial")
summary(fit.glm)


Call:
glm(formula = default ~ income + balance, family = "binomial", 
    data = Default)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.4725  -0.1444  -0.0574  -0.0211   3.7245  

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.154e+01  4.348e-01 -26.545  < 2e-16 ***
income       2.081e-05  4.985e-06   4.174 2.99e-05 ***
balance      5.647e-03  2.274e-04  24.836  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2920.6  on 9999  degrees of freedom
Residual deviance: 1579.0  on 9997  degrees of freedom
AIC: 1585

Number of Fisher Scoring iterations: 8


**Comment:**

Both income and balance are statistical signifincant (p-value < 0.05) in estimating default.

## Task (b)
Using the validation set approach, estimate the test error of this model. In order to do this, you must perform the following steps:

### I. Split the sample set into a training set and a validation set.

In [27]:
train = sample(dim(Default)[1], dim(Default)[1] / 2)

### II. Fit a multiple logistic regression model using only the training observations.

In [28]:
fit.glm = glm(default ~ income + balance, data = Default, family = "binomial", subset = train)
summary(fit.glm)


Call:
glm(formula = default ~ income + balance, family = "binomial", 
    data = Default, subset = train)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.5830  -0.1428  -0.0573  -0.0213   3.3395  

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.194e+01  6.178e-01 -19.333  < 2e-16 ***
income       3.262e-05  7.024e-06   4.644 3.41e-06 ***
balance      5.689e-03  3.158e-04  18.014  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1523.8  on 4999  degrees of freedom
Residual deviance:  803.3  on 4997  degrees of freedom
AIC: 809.3

Number of Fisher Scoring iterations: 8


### III. Obtain a prediction of default status for each individual in the validation set by computing the posterior probability of default for that individual, and classifying the individual to the “default” category if the posterior probability is greater than 0.5.

In [29]:
probs = predict(fit.glm, newdata = Default[-train, ], type = "response")
pred.glm = rep("No", length(probs))
pred.glm[probs > 0.5] = "Yes"
table(pred.glm, default[-train])

        
pred.glm   No  Yes
     No  4824  108
     Yes   19   49

### IV. Compute the validation set error, which is the fraction of the observations in the validation set that are misclassified.

In [30]:
mean(pred.glm != default[-train])

[1] 0.0254

**Comment:**
 
2.86% test error rate with the validation set approach

## Task (c)
Repeat the process in (b) three times, using three different splits of the observations into a training set and a validation set. Comment on the results obtained.

In [31]:
for (i in 1:3) {
    train = sample(dim(Default)[1], dim(Default)[1] / 2)
    fit.glm = glm(default ~ income + balance, data = Default, family = "binomial", subset = train)
    probs = predict(fit.glm, newdata = Default[-train, ], type = "response")
    pred.glm = rep("No", length(probs))
    pred.glm[probs > 0.5] = "Yes"
    print(mean(pred.glm != default[-train]))
}

[1] 0.0274
[1] 0.0244
[1] 0.0244


**Comment:**

The validation estimate of the test error rate can be variable, depending on precisely which observations are included in the training set and which observations are included in the validation set.

# Task (d)
Now consider a logistic regression model that predicts the probability of “default” using “income”, “balance”, and a dummy variable for “student”. Estimate the test error for this model using the validation set approach. Comment on whether or not including a dummy variable for “student” leads to a reduction in the test error rate.

In [32]:
train = sample(dim(Default)[1], dim(Default)[1] / 2)
fit.glm = glm(default ~ income + balance + student, data = Default, family = "binomial", subset = train)
pred.glm = rep("No", length(probs))
probs = predict(fit.glm, newdata = Default[-train, ], type = "response")
pred.glm[probs > 0.5] = "Yes"
mean(pred.glm != default[-train])

[1] 0.0278

**Comment:**

It doesn't seem that the student dummy variable can improve validation set estimate of the test error rate.

# Exercise 2
We continue to consider the use of a logistic regression model to predict the probability of “default” using “income” and “balance” on the “Default” data set. In particular, we will now computes estimates for the standard errors of the “income” and “balance” logistic regression coefficients in two different ways: 
- (1) using the bootstrap
- (2) using the standard formula for computing the standard errors in the glm() function. Do not forget to set a random seed before beginning your analysis.

## Task (a)
Using the summary() and glm() functions, determine the estimated standard errors for the coefficients associated with “income” and “balance” in a multiple logistic regression model that uses both predictors.

In [33]:
fit.glm = glm(default ~ income + balance, data = Default, family = "binomial")
summary(fit.glm)


Call:
glm(formula = default ~ income + balance, family = "binomial", 
    data = Default)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-2.4725  -0.1444  -0.0574  -0.0211   3.7245  

Coefficients:
              Estimate Std. Error z value Pr(>|z|)    
(Intercept) -1.154e+01  4.348e-01 -26.545  < 2e-16 ***
income       2.081e-05  4.985e-06   4.174 2.99e-05 ***
balance      5.647e-03  2.274e-04  24.836  < 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 2920.6  on 9999  degrees of freedom
Residual deviance: 1579.0  on 9997  degrees of freedom
AIC: 1585

Number of Fisher Scoring iterations: 8


**Comment:**

The glm() estimates of the standard errors for the coefficients β0, β1 and β2 are respectively $4.348*10^{-1}$, $4.985*10^{-6}$ and $2.274*10^{-4}$.

## Task (b)
Write a function, boot.fn(), that takes as input the “Default” data set as well as an index of the observations, and that outputs the coefficient estimates for “income” and “balance” in the multiple logistic regression model.

In [34]:
boot.fn = function(data, index) {
    return (coef(glm(default ~ income + balance, data = data, family = "binomial", subset = index)))
}

## Task (c)
Use the boot() function together with your boot.fn() function to estimate the standard errors of the logistic regression coefficients for “income” and “balance”.

In [35]:
library(boot)
boot(Default, boot.fn, 1000)


ORDINARY NONPARAMETRIC BOOTSTRAP


Call:
boot(data = Default, statistic = boot.fn, R = 1000)


Bootstrap Statistics :
         original        bias     std. error
t1* -1.154047e+01 -3.910447e-02 4.350746e-01
t2*  2.080898e-05  1.635325e-07 4.845196e-06
t3*  5.647103e-03  1.842153e-05 2.302679e-04

**Comment:**

The bootstrap estimates of the standard errors for the coefficients β0, β1 and β2 are respectively $4.344*10^{-1}$, $4.866*10^{-6}$ and $2.298*10^{-4}$.

## Task (d)
Comment on the estimated standard errors obtained using the glm() function and using your bootstrap function.

**Comment:**

The estimated standard errors obtained by the two methods are pretty close.